### ISSUE TITLE
Create separate user login for admin (restricted - webmaster only)

### ISSUE THREAD

**Issue by bsaiken** [FWAJL/FieldWorkAssistantMVC#421]
Admin user shall be able to:

PM > List All
       > Add
`Can add, edit, delete PMs.`
Link to email PMs

Project > List All  
Can Edit (view) and Delete Projects

Technicians > List All
Can edit and Delete Technicians.
Can set/unset current PM to limit Technician List.
Breadcrumb to display Current PM

Locations > List All
Can edit and delete Locations.
Can set/unset current PM and current Project to limit Location List.
Breadcrumb to display Current PM, current Project

Map > All Projects
Map > All Locations
Can set/unset current PM and current Project to limit Project and Location Lists.
Breadcrumb to display Current PM, current Project

Tasks > List All
Can edit and delete Tasks.
Can set/unset current PM and current Project to limit Task List.
Breadcrumb to display Current PM, current Project

`Analytes > Upload Common
Can Add, Edit and Delete common Analytes.`

`Field Forms > Upload 
Can Add, Edit and Delete standard form templates.`

Service Providers
Can Add, Edit and Delete recommended Service Providers.  Note - there is no table or interface for this yet.  Just being thorough.

Completed Tasks
Can view, edit delete all completed tasks.
Can set/unset current PM and current Project to limit Completed Task List.
Can package and completed forms and other Task info for delivery (email?) to PM.

**Comment by bsaiken** [comment]
Please work from the requirements we discussed on eLance.

**Comment by bsaiken** [comment]
@damirius 
I have a further clarification for this Issue.  When you are adding the login for the administrator, keep in mind that we need to have role management for 4 roles:

Project Manager (that is the role we currently have)
Administrator (this Issue)
Field Technician
Visitor
 
Each role will have a different set of privileges.  I can't go into the details for the privileges for Field Technicians right now.  I do not know the privileges for the Visitor role, or if we will actually use the visitor role.  Jeremie said that if you knew up front what the possible roles are it will be easier to set up.

**Comment by FWAJL** [comment]
@bsaiken @damirius 

As an addition, I was thinking that the role could be applied to the route and its php modules using a new attribute on the `route` and `php_module` elements called `roles` that be like the following example:

Route project/listAll:

- `role="pm"`

Route technician/showForm:

- `role="pm,ft"`

Then, if we want to show or hide a specific `php_module` based on the role, you could that too.

What do you guys think?

**Comment by bsaiken** [comment]
@damirius @FWAJL 
This Issue has vaulted to the top as there are some key SOWs I want to release, and this is a prerequisite.
This Issue really has two parts: 1) create roles and 2) create the admin interface (as described in the Jan 20 comment).  You may want to split this into two Issues.  The roles are most important, because we want to release the SOW for the Field Technician UI.

**Comment by damirius** [comment]
@bsaiken @FWAJL 

Permission system can be one of the hardest things to implement. This is a very serious Issue. As far as I can see, roles should be used to mark in routes.xml file which modules and routes can be accessed. I'm sure that restricting the user from accessing some modules on the page but giving him permission to see the page will result in some serious issues, but I guess we will see soon enough.

**Comment by FWAJL** [comment]
@bsaiken @damirius 

I'm going to help you do it, even start it. I have the database tables update almost ready.
Let's just try and see. We don't need to implement a too complex system.

More soon!

**Comment by FWAJL** [comment]
@damirius 

Have a look at the script here: https://github.com/FWAJL/FieldWorkAssistantMVC/blob/dev/Database/Scripts/Updates/2015-03-19.sql

It is a draft of the database structure. Please let me know if you have suggestions.
I think I will identify more accurately the tasks for this so we can split them into small dev items and then we will start.

Thank you.

**Comment by damirius** [comment]
@FWAJL 

`user_hint` contains some hint for password right?
Can you tell me how do you plan to use `user_is_logged`?

Administrator role is interesting one, since users of that role should be able to "login" as other users, not including other admins of course. So it's kinda special role.

I'll share here my thoughts on this that I already mentioned to Brian:

It would be great to split the Left menu into two parts for admins. One part would be admin only menu with few items.
Let's say:
Manage users (PM) - list of all project managers, admin can right click any of them to choose to edit them, delete them or log in as them
Manage common analytes - this will show two tabs with the lists for common field analytes and common lab analytes, admin can edit them
Manage common forms - list of existing common forms and functionality to add new ones

When logged in admin takes the role of some project manager he will get the second part of the menu which is actually just existing menu we already use for regular PM. In that case, we would avoid too much modifications of existing controllers/pages.

**Comment by bsaiken** [comment]
@damirius @FWAJL 
I can answer some of these.
1) Yes, `user_hint` will contain a password hint.  It is not used right now, but eventually we will incorporate it.  Once the site is up and running, we will "encourage" users to have complex passwords, as a security feature.  In order  to market this website to my target audience, security will have to be very high.  If any user's information became compromised, it would doom the site.
2) We will use `user_is_logged` to satisfy Issue #250.  Once a Project Manager is logged in, no-one else can log in with that user id.  Later we will add a timeout so no activity for a period of time results in the PM being logged out automatically.  It's all part of the work toward beefed up security.
3) Obviously, the other items we discussed and I like your approach.

**Comment by damirius** [comment]
@bsaiken @FWAJL 

Maybe it's better to store some timestamp there and to update it every time user access the page? Since we must enable the user to login again after some period of no activity. I am aware that we have `log` table we can add `user_id` there and check it like that if we want to do it that way. One important thing is to set that "timeout" time to smaller value than session timeout. If user's session runs out he will not be able to see the pages and he will have to log in again and if our `user_is_logged` field is still set to 1, he will not be able to do so till some script clears it.

**Comment by bsaiken** [comment]
@damirius @FWAJL 
Maybe this is not needed if there is a session timeout - it seems redundant.  
The condition I want to cover is this:  
A Project Manager logs in (from their desk at work) and then needs to leave the office and go home.  Then, at home, they need to log in again and continue working.  I don't want them locked out.

Another thing to keep in mind is that I will want to address site security as a completely separate, and comprehensive Issue, so don't let me get distracted with ideas that will become irrelevant.

I thought it might be handy during development because I sometimes log in as "demo", or some other user, to set up some demonstration conditions, and I don't want someone else logging in under the same user_id at the same time.

**Comment by FWAJL** [comment]
@bsaiken @damirius 

Just an article on session expiration to look at: http://stackoverflow.com/questions/8311320/how-to-change-the-session-timeout-in-php

I will comment more soon.

Thanks.

**Comment by bsaiken** [comment]
@FWAJL @damirius 

This looks a bit complex, but it would be safe to assume that our clients will be not malicious, but I did not think we were using cookies.

There are numerous sites that I use that automatically log me out after a period of inactivity, so there must be a common solution.

**Comment by FWAJL** [comment]
@bsaiken @damirius 

Only the following what we would use (not the cookies) in regards to session timeout management:

```
session_start(); // ready to go!

$now = time();
if (isset($_SESSION['discard_after']) && $now > $_SESSION['discard_after']) {
    // this session has worn out its welcome; kill it and start a brand new one
    session_unset();
    session_destroy();
    session_start();
}
```
That is to implement a session timeout that we can configure.

**Comment by damirius** [comment]
@bsaiken @FWAJL 

We would need this line too `$_SESSION['discard_after'] = $now + 3600;` if we want to refresh it when users makes new request. If we don't do that user will be logged out after 1 hour even if he was active for the whole hour. It's a nice trick to keep clearing the session exactly when needed. That number of seconds can be set to `ini_get("session.gc_maxlifetime");` if we wish, so that we use exact session timeout time. On the other hand it doesn't logs out the user from the other browser/machine. I think that storing the `session_id` in the table instead of `user_logged_in` would be an OK solution. We can always check if the user is logged in and if the user have same `session_id` as the one stored in db for him. When the user change browsers/machines and log in, new `session_id` will be stored there and everything will continue as normal. If he tries to access other pages in the old browser/machine script will check for his session_id and since new session_id is stored in the db it will prevent him from accessing pages. He can go on login page and login again.
That said, I think that we need to implement this aggressive removal of sessions plus this thing with storing session_id in db.

Also I want to note that sessions by default have their IDs stored in cookies. Everything in sessions is stored on the server of course, but ID is stored in the user's cookie also so that server knows which session data belongs to which user. So yes, if someone can access cookie files on the users machine he can steal session_id and set it in his cookie and login as that user. Other popular thing to do is XSS, inject some malicious js code into some unsanitized input box. I.E. comment section with the JS and read his session_id cookie for that website, then redirect that user to your script hosted somewhere else with his session_id in the query variable. Keep in mind this is sorted out mostly by setting httponly flag which prevents the js to access all cookies like this one, so this isn't a problem anymore.
In our case if we store the session_id in the db we can hash it before storing and check it like we check passwords by hashing it before the check.

@FWAJL
I wanted to ask you about one thing that's bugging me with the Admin role. Since it was decided that Admin can "log in" as some other user (that's not admin too) and access all pages they can and modify data like them I ran into some little problem. First off I want to say that I find this solution rather easy to implement since if we let him to log in as the other user we don't need to add special fields/buttons for him on every page. He can just log in as someone else and that's it everything is already there. Note that by logging in I don't mean for him to use their user/pass combo, he will have some special page "manage users" where we will have all users listed for him and he can click something like "log in as this user". The problem that I have is when he wants to revert back to his admin role. Since when he logs in as the regular user (pm/technician/etc) he will see everything like they do, but I need to show him somewhere link to log back in as Admin. I don't think that they need to go back to login page and type again their user and pass. I was thinking of storing something else in session which will be used as a flag to mark them as the admin even when they are working on some other account. What do you think?

**Comment by bsaiken** [comment]
@damirius @FWAJL 

I can answer this question: 
*The problem that I have is when he wants to revert back to his admin role*

Once an admin logs in as another user, they would need to log out and log in again as admin.

**Comment by damirius** [comment]
@bsaiken @FWAJL 

Great! This solves some problems

**Comment by FWAJL** [comment]
@damirius @bsaiken 

I agree to store the session id as the value to check if the user is logged or not. This will be issue #745.

As for the admin being able to view a PM data, this doesn't require "logging in" as that selected PM. Instead, when the admin logs in, we will create an array of PM Session arrays. The Admin selects the PM he wants to view and then the rest works just as if he was the PM because we use the `PmHelper::GetCurrentSessionPm()` everywhere in the application.

On the UI, the Admin would see the same left menu as the PMs at the exception of the Project Managers list and some other items like the CommonAnalytes and so on.

So no need to login / logout the Admin.

However, to edit any data, the Admin ***must select*** a PM. This is to keep it simple.

Is it all clear?

**Comment by FWAJL** [comment]
@bsaiken @damirius 

> On the UI, the Admin would see the same left menu as the PMs at the exception of the Project Managers list and some other items like the CommonAnalytes and so on.

By that, I mean we can later apply the role Admin to the routes we want him to see.

**Comment by damirius** [comment]
@bsaiken @FWAJL 

I'm working on this. Since we have roles now and different user types it gets a little tricky since everything is based on assumption that user is project manager and that's no longer correct.

**Comment by FWAJL** [comment]
@damirius @bsaiken 

Have you retrieved the database updates with the tables user and user_role? 
I think we also need to split this task into smaller ones. Sorry I only say this now...

Here is what I think we have to do:

1. Modify authentication to use the user table so that we can still login with the exisiting project managers. This means to copy the data from the project_manager table to the user table so that the credentials are intact. No role logic for now.
2. Retrieve the role attribute on the route element in routes.xml to store it as an array of roles in the Route object. After thinking about it, it'll be easier to create a new route (and its controller method) for the FT interface rather than trying to PM and FT logic in the controller method (same route). 
3. Filter the routes list for the user logging in to contains only the routes he can view. Include a fallback route so that the user sees a message when he tries to access a page he can't view (this will be useful to find bugs in development)
4. Build the PM list UI: will add a separate issue for that.

Where are you in developing these points?

Thank you.

**Comment by FWAJL** [comment]
@damirius 

I have updated the comment above.

**Comment by damirius** [comment]
@FWAJL

Thanks. I already finished the new login, but I want to finish up security layer classes and move the code there, so I'm currently working on that.

**Comment by FWAJL** [comment]
@damirius 

Have you done the update to store the session id? If so, take the issue #745.

Thank you.

**Comment by damirius** [comment]
@bsaiken @FWAJL 

I've finished this first part and made some Auth classes. **UPDATE YOUR DB WITH 2015-03-25.sql SCRIPT**
Since logging in will use data from new table, if you don't populate it properly you will not be able to log in.
From the front end it should look the same as before.

**Comment by FWAJL** [comment]
@damirius @bsaiken 

Your script was missing the backward compatibility, Damir. In order to release the update to the live, always think about the data and how we want to keep its integrity. 

So you want to create an INSERT statement that does what my updates show you. It copies the data and create an INSERT that will work on any server, your, mine, Brian's and Souvik's. 

I'll let Brian do the function review before looking in depth in your code. So far, it works :+1: 

Thank you.

**Comment by bsaiken** [comment]
@damirius @FWAJL 
I need to step in and see where we are at before going any further.  So, let's table work on this Issue for the moment.

I do want to review commit 6b3afbf.  However, as far as a function review, it is not clear to me what front-end functionality I am looking for.

**Comment by FWAJL** [comment]
@bsaiken 

Run all your unit test as you should always do and the frontend functionality will be reviewed.

Thanks.

**Comment by bsaiken** [comment]
@FWAJL 

There is a problem with this request.  I do function reviews on my local computer.  In doing so, I can review any feature branch.  If I need to, I can update my local database using the method I mentioned earlier (downloading the database from the live site since the scripts don't work).  

Unit testing is different.  I do unit testing on the live site once the code has been released.  Are you telling me you want me to do unit testing on my local machine using the feature_damir branch?

**Comment by bsaiken** [comment]
@FWAJL 
I did some limited testing on my local computer with the feature_damir branch and the updated database tables.  It seems to work just fine.  :-)  It was not a thorough unit test.  If you need me to do that let me know, but be specific about what testing you need.

As a side note, I spent (OK, wasted) a fair amount of time trying to test the site with IE.  My sister has a very old computer running IE7.  There are many bugs with IE7.  We don't need backward compatibility - I'm just telling you for future reference in case you go crazy and try testing with IE7.

**Comment by FWAJL** [comment]
@bsaiken 

By backward compatibility mentioned previously, I meant in relation to the old database and the newest (after the updates made on March 19th, 23rd and 25th).

Unit testing is not limited to the live site. In fact, you use most of the test to validate the most important points (login, add a project, add a service, etc...). This is to avoid releasing code bugs. Do you understand?

IE7, 8 and even maybe 9 are a nightmare to work with. We need to focus on most recent browsers because it goes in the same direction as your idea for this website: use the latest technologies. Trying to use IE7 to 9 would be like sticking to paper forms.

Thank you for reading.

**Comment by FWAJL** [comment]
@bsaiken @damirius 

I think that in order to answer Brian's need described in #640, we still need to implement the roles and to do that, we just need to steps 2 and 3 described above (issues #747 and #748). Then the Mobile integration will start with #748 since you need to test at least with a PM and a FT.

The steps 4 to 6 can be done later, of course.

@bsaiken : does it make sense and are you ok to proceed that way?

Thank you.

**Comment by bsaiken** [comment]
@FWAJL @damirius 
I do think I am understanding the approach.  I think Item 2, above, addresses my concerns.  As you wrote in your comments to Issue #640, we will need a completely different set of views and controller methods for the FT site.  

As I considered further the details of the UI needed for the FTs, I realized that even though some of the pages appear similar to the PM pages, they are, in fact, very different.

**Comment by FWAJL** [comment]
@bsaiken @damirius 

So can Damir proceed with #747 and #748? Thank you.

**Comment by bsaiken** [comment]
@damirius @FWAJL 

Damir, 

This issue has become the highest priority because I need to be able to properly upload master forms.  We manually put some master forms in the uploads/master_forms directory, but without the relation table completed, it is delaying other work.  
For the admin user, I have temporarily limited the requirements to 

`Analytes > Upload Common
Can Add, Edit and Delete common Analytes.`

and

`Field Forms > Upload 
Can Add, Edit and Delete standard form templates.`

in order to clear some bottlenecks.  

It would be helpful to have admin and field technician users to work with, so could you take on Issue #809 as well?

**Comment by damirius** [comment]
@bsaiken @FWAJL 

I've added some basic redirect based on user role, so that admins go directly to analyte/uploadCommons instead to project list.
Since we have now different type of users, we can't use existing create user method we used before in project manager controller.
For testing, **please insert new user in the db**
For example, use "admin" as the `user_login` and copy the password from some other user, for example from "test" user, so you'll have user:pass combo admin:test. Set the `user_role_id` field to 1, since 1 is the admin role.
When you login with new admin user, you'll have new menu with two pages, one is for adding new template forms and other is for adding common analytes. I found that page for uploading analytes was already there so I used it like that. 

Keep in mind that for now you can only ADD new forms and/or analytes. I wanted to ask you how to organize all those functions. We already have upload/delete for user forms, we can do it like that for master forms too. Keep in mind that there is no point in "editing" forms since they are pdfs, we can just remove them and upload another one if there aren't any issues with that.

Common analytes are a different story. I see that we used text box for them so we can insert multiple at once. Where and how to list existing ones? Also I noticed that lab analytes have some category name, I don't know where to specify that in the box, so for now category name is the same as the analyte name till we find the solution for this.

Thanks

**Comment by bsaiken** [comment]
@damirius  @FWAJL 

Damir,

I understand about the logins for other users.  I will review later today.  

Would you be able to take on Issue #809, in order to resolve this?  See Issue #809 for additional comments.

For the admin login, we can allow the admin user to add and delete master forms and common analytes.  Right now we can't do either, because we have not had an admin login. 

To be honest, I am not satisfied with the way I have set up the Analytes interface.  So, let's just leave the analytes alone for now.  The real bottleneck is uploading master forms...

This Issue (#421) has a lot of potential functions.  Therefore, I will create separate Issues for each new Admin function as the need arises.  The next Issue I will create is giving the admin the ability to create a new PM user.  However, we will need Issue #809 finished before we do that.

**Comment by damirius** [comment]
@bsaiken @FWAJL 

Hm, I thought that Issue #809 is one that takes care of creating new users by admin.
Also, you can upload master forms now, you just need to login as admin. I'll see that I add list of all master forms for admin to view and edit also. For now you can only upload new ones

**Comment by bsaiken** [comment]
Well, your right. But we do need to create one admin user.   If I understand you correctly, we could just create a PM user, the edit the database table and change the role to admin, and that would do it, right?

**Comment by bsaiken** [comment]
@damirius @FWAJL 

What I intended for Issue #809, is that we write a method that allows us to create users with different roles.  The method is generic, but the implementation depends on when the method is called.  I will give details in Issue #809.

Jeremie had written a method to create new users, but they were always PM users, and it was before we implemented the roles. Now there is no way to create a new user right now.

**Comment by bsaiken** [comment]
@damirius 

Damir,

I tried setting up a new user login "admin", but I must not have understood your instructions because it did not work.  Can you please be more specific so I can set this up on the server.

**Comment by FWAJL** [comment]
@bsaiken
 
Use this script to add your admin: https://github.com/FWAJL/FieldWorkAssistantMVC/blob/feature_damir/Database/Scripts/Updates/2015-04-06.sql

**Comment by bsaiken** [comment]
@FWAJL  @damirius 
OK - this is working and we have an admin user.  I ran the script on the Hurray Host database - so we should have an active admin server once the code is merged.
I was successful in uploading a master form.
I successfully upload common lab and field analytes.
I will work tomorrow on creating some new Issues to complete the Admin user interface and updating the function test spreadsheet.
Damir - please let me know if you have access yet to the unit test spreadsheet - I thought I sent you an updated link.

**Comment by damirius** [comment]
@bsaiken @FWAJL 

I have access to the unit test spreadsheet. This one here https://docs.google.com/spreadsheets/d/1lNhCFAW1PXiS6QDUDzWfAAbJG5VTZr-jQchaeWhRgXk
I've added sorting tests for each issue yesterday. I'll add new ones for this too, didn't want to do it then since I didn't know if we are going to split this into more issues or not.

**Comment by bsaiken** [comment]
@damirius @FWAJL 
I will close this Issue as the admin user interface has been created.  Any further enhancements and changes will be prepared as separate Issues.

Brian
